### ISSUE TITLE
Pretrained weights & freeze_feature_extractor arguments for ResNet

### ISSUE THREAD

**Issue by rhine3** [kitzeslab/opensoundscape#328]
Noticed that the InceptionV3 CNN architecture classes each have a `use_pretrained` argument to turn on ImageNet pretrained weights:

https://github.com/kitzeslab/opensoundscape/blob/58eee0dfe473ee73f53bd615773c88f493e2c52e/opensoundscape/torch/models/cnn.py#L938

whereas the ResNet18 architecture classes do not and pretrained weights are always loaded:

https://github.com/kitzeslab/opensoundscape/blob/58eee0dfe473ee73f53bd615773c88f493e2c52e/opensoundscape/torch/models/cnn.py#L806

Default behavior should definitely be to use pretrained weights, but shouldn't both of these classes have the same option to turn off pretrained weights without going down into the `resnet` function? I noticed that the ResNetArchitecture class docstring does say that `weights_init = None` uses a random weight initialization, but it looks like that behavior isn't implemented yet:

https://github.com/kitzeslab/opensoundscape/blob/58eee0dfe473ee73f53bd615773c88f493e2c52e/opensoundscape/torch/architectures/resnet.py#L63-L71

Edit: Same thing goes for freeze_feature_extractor (the ability to turn this on would be even more useful than turning off pretrained weights, IMO)

**Comment by sammlapp** [comment]
Good catch on the not implemented part. So what happened is the ResNet18Binary and Multiclass are based on the original classes Miao wrote, but are a bit more “custom” than the most generic opso classes. The PytorchModel class is the generic class that can take any architecture, including resnet, and the Inception classes are that class adapted for inception. Those are the classes that fit the opso framework and structure, and we should maintain. The Resnet18… classes are kind of legacy or custom in that they use an internally implemented version of the resnet architecture. Thats why they don’t have the same initialization structure, and are created a bit differently. Maybe they should be renamed and tutorials should not use them. But they are what we’ve mostly been using to train models, because they allow the added flexibility of separate training parameters for feature and classification layers. (Enabling this generally would require us to have internal implantation of every architecture). We can discuss more.

**Comment by rhine3** [comment]
Ah ok so correct me if I'm wrong:
* `ResNetArchitecture` is an architecture based on Miao's previous implementation and allows different training params for feature extraction layers and classifier layers
* `Resnet18...` legacy classes use this architecture
* PytorchModel class is an entirely separate paradigm from the `Resnet18...` classes and is our "modern" framework
* In practice we could freeze the feature layer while using `Resnet18...` classes by setting the feature learning rate to 0?

It sounds like we could technically implement the random weight initialization, but because the `Resnet18...` classes are more of a lab-internal feature and we rarely use random weight initialization, this is probably unnecessary. Maybe we can just add a `NotImplementedError` to handle the `weights_init=None` argument in case we did ever want to implement that feature.

So regarding the docs: do you think that in most cases, someone wanting to train a ResNet model (including us and our collaborators) would be just as happy using the `PytorchModel` framework, or do you think that the `ResNet18...` classes provide a significant advantage?

**Comment by sammlapp** [comment]
Almost all correct. `PytorchModel` is actually the parent of all of these, and the infrastructure is almost all shared between all of the classes. The Resnet18Multiclass and Resnet18Binary classes just supply a custom architecture rather than one of the "stock" pytorch architectures, and add a couple of functionalities. (And yes, you could technically set the learning rate to 0 for feature block, but you'd still be wasting time computing gradients for all of those, so it would be better to set torch.no_grad() for all of the network parameters you don'w want to train, much like https://github.com/kitzeslab/opensoundscape/blob/58eee0dfe473ee73f53bd615773c88f493e2c52e/opensoundscape/torch/architectures/cnn_architectures.py#L36 

I'm not sure which one would be best to use in tutorials/docs - I'm inclined to use the more generic PytorchModel framework because it is more aligned with the opso framework and will be more forward-compatible; but, I do think we should continue to use ResNet18 models internally for now until we know if performance is comparable. (We do know the more generic models work pretty darn well based on the architecture-comparison tests I ran).

**Comment by sammlapp** [comment]
rewrote classes to subclass PytorchModel, this should be better / clearer / more consistent now
